### PR TITLE
Add support for Python 3.12

### LINKED ISSUE #10887: Support for Python 3.12
This issue tracks various components that need to happen before we can be said to fully support Python 3.12.  Notably, we will _automatically_ have wheels available for Python 3.12 already, as since #10120 we've built our wheels against the stable Python C API, which means our released wheels should already be ABI compatible with Python 3.12.  However, not all our dependencies will have pre-built wheels for Python 3.12.

```[tasklist]
### Binary dependencies need to release 3.12 wheels
- [x] Numpy (as of 1.26.0)
- [x] Scipy (as of 1.11.2)
- [x] psutil (as of 5.9.4 it's released abi3 wheels)
- [x] Rustworkx (Qiskit/rustworkx#986)
- [x] Symengine (https://github.com/symengine/symengine-wheels/issues/11)
```

```[tasklist]
### Developer dependencies with problems
- [x] `testtools` fails to import due to the removal of `distutils`.  Should be fixed in next release. (testing-cabal/testtools#352, testing-cabal/testtools#353)
- [x] `stestr` complains that "run is not an 'stestr' command" with Python 3.12.  This turned out to be a knock-on effect of the `testtools` problem above.
- [x] `python-subunit` uses deprecated `datetime.datetime.utcnow`, issuing warnings that fail our suite during collection.  We can add skips for those warnings if necessary.  See testing-cabal/subunit#71.
```

```[tasklist]
### Internal to Qiskit
- [ ] We have problems with Numpy 1.25+ in some isometry code, but will need to support 1.26 for Python 3.12.  See #10305, numpy/numpy#24000.
- [ ] We have problems with Scipy 1.11.0+ on Windows, most likely pre-existing.  See #10345.
- [x] Use of deprecated `ast` features. See #10886
```

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

Python 3.12.0 was released on 10-02-2023, this commit marks the start of support for Python 3.12 in Qiskit. It adds the supported Pythonv ersion in the package metadata and updates the CI configuration to run test jobs on Python 3.12 and build Python 3.12 wheels on release.

### Details and comments

Fixes: #10887

### CODE REVIEW

**Comment by mtreinish** [comment]
I think this should be good to go now, at least it passes locally. My only concern is we'll hit the same stability issue with unitary synthesis on the newer numpy that led to #11232. But to support 3.12 we need to use a newer version of numpy.

File: qiskit/providers/fake_provider/fake_backend_v2.py [hunk 1]
```diff
@@ -42,7 +42,7 @@ def __init__(self):
             None,
             name="FakeV2",
             description="A fake BackendV2 example",
-            online_date=datetime.datetime.utcnow(),
+            online_date=datetime.datetime.now(datetime.timezone.utc),
```
Comment (jakelishman) [inline qiskit/providers/fake_provider/fake_backend_v2.py:45]: I was confused why I hadn't seen and fixed this when I was doing the initial Python 3.12 testing, but then I remembered that symengine didn't have wheels at the time and we had Numpy/Scipy problems, so I'd not got all the way.

**Review by jakelishman — DISMISSED** [review]
This seems fine by me.  We ought to get to the bottom of the Numpy flakiness - the failing test is weird to me full stop, but I'm most concerned that it's flaky.

**Comment by mtreinish** [comment]
I've been rerunning the tests and managed to trigger the test failure on the 3rd run: https://dev.azure.com/qiskit-ci/qiskit/_build/results?buildId=51680&view=logs&j=783cd8ce-45d0-52b9-fbe2-d8b6f5c36b94&t=f6ef7ac4-a492-580b-ed69-1d6c3306a466&l=22108 I think we need to get to the bottom of this instability issue before we can merge this, otherwise we'll likely block ci in practical terms because of these failures.

**Comment by jakelishman** [comment]
Unblocked by the merge of #11307.

**Review by jakelishman — APPROVED** [review]

**Pull request merged**